File: hw1/program/container.cpp
//------------------------------------------------------------------------------
// container.cpp - содержит функции обработки контейнера
//------------------------------------------------------------------------------

#define _CRT_SECURE_NO_WARNINGS
#include "stdio.h"
#include "container.h"

//------------------------------------------------------------------------------
// Инициализация контейнера
void Init(container& c) {
    c.len = 0;
}

//------------------------------------------------------------------------------
// Очистка контейнера от элементов (освобождение памяти)
void Clear(container& c) {
    for (int i = 0; i < c.len; i++) {
        delete c.cont[i];
    }
    c.len = 0;
}

//------------------------------------------------------------------------------
// Ввод содержимого контейнера из указанного потока
void In(container& c, FILE* file) {
    int k;
    while (fscanf(file, "%d", &k) != EOF) {
        if ((c.cont[c.len] = In(file, k)) != 0) {
            c.len++;
        }
    }
}

//------------------------------------------------------------------------------
// Случайный ввод содержимого контейнера
void InRnd(container& c, int size) {
    while (c.len < size) {
        if ((c.cont[c.len] = InRnd()) != nullptr) {
            c.len++;
        }
    }
}

//------------------------------------------------------------------------------
// Вывод содержимого контейнера в указанный поток
void Out(container& c, FILE* file) {
    fprintf(file, "Container contains %d elements.\n", c.len);
    for (int i = 0; i < c.len; i++) {
        fprintf(file, "%d: ", i + 1);
        Out(*(c.cont[i]), file);
    }
}

//------------------------------------------------------------------------------
// Вывод обработки данных контейнера
void ProcessingVar21(container& c, FILE* file) {
    double sum = 0.0;
    for (int i = 0; i < c.len; i++) {
       sum += SurfaceArea(*(c.cont[i]));
    }
    double mean = sum / c.len;
    fprintf(file, "Arithmetic mean: %f\n", mean);

    int k = 0;
    fprintf(file, "Surface areas processed:\n");
    for (int i = 0; i < c.len; i++) {
        if (SurfaceArea(*(c.cont[i])) >= mean) {
            k++;
            fprintf(file, "%d. ", k);
            Out(*(c.cont[i]), file);
            //fprintf(file, "%d. %f   ", SurfaceArea(*(c.cont[i])));
        }
    }

}

File: hw3/program/container.py
from sphere import *
from tetrahedron import *
from parallelepiped import *
# ----------------------------------------------


class Container:
    def __init__(self):
        self.store = []

    def get_figures_from_file(self, data):
        i = 0
        shape = 0
        while i < len(data):
            fig = int(data[i])
            if fig == 1:
                shape = Parallelepiped()
                shape.get_figure_from_file(data, i+1)
                i += 3
            elif fig == 2:
                shape = Sphere()
                shape.get_figure_from_file(data, i+1)
                i += 1
            elif fig == 3:
                shape = Tetrahedron()
                shape.get_figure_from_file(data, i+1)
                i += 1
            i += 2
            self.store.append(shape)

    def get_random_figures(self, count):
        i = 0
        shape = 0
        while i < count:
            fig = random.randint(1, 4)
            if fig == 1:
                shape = Parallelepiped()
                shape.get_random_figure()
            if fig == 2:
                shape = Sphere()
                shape.get_random_figure()
            if fig == 3:
                shape = Tetrahedron()
                shape.get_random_figure()
            i += 1
            self.store.append(shape)

    def write(self, output):
        output.write("Container is store {} shapes:\n"
                     .format(len(self.store)))
        for shape in self.store:
            shape.write(output)
            output.write("\n")

    def processing_var_21(self, output):
        sum = 0.0
        for shape in self.store:
            sum += shape.surface_area()
        mean = sum / len(self.store)
        output.write("Arithmetic mean: {}\n".format(mean))
        output.write("Surface areas processed:\n")
        for shape in self.store:
            if shape.surface_area() >= mean:
                shape.write(output)
                output.write("\n")
        output.write("\n")
